components/uuf-core/src/main/java/org/wso2/carbon/uuf/internal/deployment/OsgiPluginProvider.java

```diff
@@ -63,27 +64,53 @@ protected void deactivate(BundleContext bundleContext) {
     @Override
     public <T> T getPluginInstance(Class<T> type, String className, ClassLoader classLoader)
             throws PluginLoadingException {
-        T plugin = loadOsgiPlugin(type, className);
-        return (plugin != null) ? plugin : loadNonOsgiPlugin(type, className, classLoader);
+        if (bundleContext == null) {
+            return loadNonOsgiPlugin(type, className, classLoader);
+        }
+        return loadOsgiPlugin(type, className);
     }
 
+    /**
+     * Creates an instance of the specified class type in OSGi mode.
+     * This method will throw {@link PluginLoadingException} when the relevant bundle is not found or on any other
+     * exception
+     *
+     * @param type      type of the instance to be created
+     * @param className name of the class
+     * @param <T>       type of the instance to be created
+     * @return instance of the specified class type
+     */
     private <T> T loadOsgiPlugin(Class<T> type, String className) {
         if (bundleContext == null) {
-            return null;
+            throw new PluginLoadingException("Bundle context is null");
         }
-
-        ServiceTracker<T, T> serviceTracker = getServiceTracker(type);
-        @SuppressWarnings("unchecked")
-        T[] services = serviceTracker.getServices((T[]) new Object[serviceTracker.size()]);
-        for (T service : services) {
-            if ((service != null) && (service.getClass().getName().equals(className))) {
-                return service;
+        Class<?> clazz = null;
+        for (Bundle bundle : bundleContext.getBundles()) {
+            try {
+                clazz = bundle.loadClass(className);
+            } catch (ClassNotFoundException e) {
+                // This bundle doesn't have the class
             }
         }
-
-        return null;
+        if (clazz == null) {
+            throw new PluginLoadingException("Class " + className + " do not exist in any bundle");
+        }
+        try {
+            return type.cast(clazz.newInstance());
```

Comment:
Here we are instantiating a new instance. This is **NOT** the intention of the OSGi plugin provider. `OsgiPluginProvider` supposed to instances of OSGi services which are already available in the OSGi environment. Creating a new instance is not the correct way.